Add cookbook examples on error handling (#5538)
dagger/dagger: docs/current/742989-cookbook.md

[Learn more](./guides/723462-use-secrets.md)

## Error handling

### Terminate gracefully

The following code listing demonstrates how to handle errors gracefully, without crashing the program or script running Dagger pipelines.

```go file=./cookbook/snippets/error-handling/aborting/main.go
```

### Handle exit code and unexpected errors

The following code listing demonstrates how to handle a non-zero exit code (an error from running a command) in a container, with several use cases:

> - Difference between “test failed” and “failed to test”

Reviewer: ```suggestion
- Handle differently “test failure” and “failure to test”
```

- Handle a specific exit code value
- Handle a failure from a command executed in a container, without checking for the exit code

> - Catching and handling a failure from a command executed in a container, without propagating it

Reviewer: ```suggestion
- Catch and handle a failure from a command executed in a container, without propagating it
```

[Learn more](./guides/723462-use-secrets.md)

## Error handling

> ### Aborting with unhandled error

Reviewer: As we discussed in Discord: I think we can skip this recipe as it merely shows the default situation anyway ie. a script without error handling. It's better to show a recipe that actually includes error handling (which I guess is the next one), so the user can copy the approach. It's also a bit odd to me that the section is called "Error handling" but the first recipe is "abort with unhandled error".

> ### Aborting with unhandled error

Reviewer: ```suggestion
### Abort pipeline with unhandled error
```

Author: Maybe just "Abort with unhandled error"?

Reviewer: Modified

Reviewer: is `Abort pipeline script with unhandled error` clearer?

Author: I was hoping abort was clear enough. Abort means quit with an error.

I'm ok with the change, but the word "pipeline" is loaded and we're still working out the terminology. While script is accurate in Python and Node.js, it's only that in Go when doing `go run`, but it's usually considered an app or program.That's why I just put "Abort", to avoid having to name things 😛

Author: FTR, in Go:

> Exit causes the current program to exit with the given status code. Conventionally, code zero indicates success, non-zero an error. The program terminates immediately; [...]

\- [os.Exit](https://pkg.go.dev/os#Exit)

Author: Going to remove.

> The following code listing demonstrates how to abort the script with an unhandled error, which allows debugging with the stack trace of the crash during development.

Reviewer: ```suggestion
The following code listing demonstrates how to abort a Dagger pipeline with an unhandled error, which allows debugging with the stack trace of the crash during development.
```

Reviewer: See if my mod is better?

Author: Going to remove.

```go file=./cookbook/snippets/error-handling/unhandled/main.go
```

> ### Aborting gracefully

Reviewer: ```suggestion
### Abort pipeline gracefully
```

Author: Same, the meaning here is for "exiting" the "program".

Reviewer: Isn't the "program" here specifically a Dagger pipeline?

Author: I meant "program" as the general concept here not a word suggestion. `go run` runs as a script, `go build` compiles in a program.

I don't see the script as the pipeline, I see the script as containing a pipeline. In this case the script only has a pipeline but I still don't equate the two. Our CI code using mage is more a script with multiple pipelines (one for each entrypoint).

Reviewer: OK I see. Is this recipe explaining how to abort a pipeline (which is within a script) gracefully and then continue to run the remainder of the script, or abort the entire script (including one or more pipelines contained within it) gracefully?

Author: abort = exit with error

The use case here is to not let a pipeline error crash the program and instead handle it and terminate cleanly. 

A crash is abrupt, may not clean up system resources and displays an ugly stack trace to the user. Aborting with a "panic" is quick and easy, but this recipe shows how to do it cleanly.

Author: I changed it to "Terminate" and "program or script".

Author: I'm going to resolve this line since it's outdated, add new comment for follow-up feedback.

> The following code listing demonstrates how to abort the script without crashing.

Reviewer: ```suggestion
The following code listing demonstrates how to abort a Dagger pipeline gracefully.
```

Reviewer: See if better?

Author: Changes the meaning. It's the program we want to abort (or terminate) cleanly and return to the shell (or whatever user interface), not the pipeline. When the pipeline fails, there's no more pipelines to run, so no other pipeline to abort (abort = don't run what comes next, just leave now).

```go file=./cookbook/snippets/error-handling/aborting/main.go
```

> ### Exit code and unexpected errors

Reviewer: ```suggestion
### Handle exit codes and unexpected errors
```

Author: "exit codes" (plural) reads funny to me. The use case is to handle an exit code (singular), it just so happens that the snippet demonstrates multiple ways you can do that.

Reviewer: Modified

> The following code listing demonstrates how to handle a non-zero exit code from running a command, with several use cases:

Reviewer: ```suggestion
The following code listing demonstrates how to handle a non-zero exit code or an error from running a command in a container, with several use cases:
```

Author: "non-zero exit code or an error from running a command in a container" reads ambiguously as either two possibilities or a clarification of same thing using different words. The intent is the latter, if you think that's clear than I'm 👍

Reviewer: See if clearer now?

> The following code listing demonstrates how to handle a non-zero exit code from running a command, with several use cases:

Reviewer: I found this example quite difficult to grok because it covers multiple use cases. I recommend splitting this into a separate recipe for each use case for readability.

Reviewer: Yes, OK to merge as is and split into separate snippets later. Added a note for this in https://github.com/dagger/dagger/issues/5539#issuecomment-1662082694

The following code listing demonstrates how to handle a non-zero exit code from running a command, with several use cases:
- Difference between “test failed” and “failed to test”
- Handle a specific exit code value

> - Handle an exec error without checking for the exit code

Reviewer: ```suggestion
- Handle an error without checking for the exit code
```

Author: For clarity, the point of "exec" here was to distinguish from other types of errors that don't have an exit code. So when you do this:

```go
if err != nil {
    var e *dagger.ExecError
    if errors.As(err, &e) {
        // exec error, handled and not propagated
        return nil
    }
    // non-exec error, propagated
    return err
}
```
But removing "exec" doesn't make it inaccurate necessarily, so it's ok.

Author: Since in my next comment removing "exec" loses meaning, if you have another suggestion, keep this one consistent with it as well.

Reviewer: See if mod makes it clearer?

> - Handle an exec error without propagating

Reviewer: ```suggestion
- Handle an error without propagating it
```

Author: Here it loses a bit more meaning. The use case is to not propagate an ExecError, but let other errors propagate.

Reviewer: See if mod makes it clearer?

> - Get the stdout of an exec, whether it failed or not

Reviewer: ```suggestion
- Get the standard output of a command, irrespective of whether or not it failed
```

```go file=./cookbook/snippets/error-handling/exit-code/main.go
```

> ### Running container after a failed exec

Reviewer: ```suggestion
### Continue using container after a command execution fails
```

> This code listing demonstrates how to continue using a container after a failed exec. A common use case for this is to export a report that a test suite tool generates.

Reviewer: ```suggestion
This code listing demonstrates how to continue using a container after a command executed within it fails. A common use case for this is to export a report that a test suite tool generates.
```

> **Important note:** The caveat with this approach is that forcing a zero exit code on a failure caches the failure. Sometimes that’s ok, but it may not be, so it requires discernment for each use case.

Reviewer: ```suggestion
:::note
The caveat with this approach is that forcing a zero exit code on a failure caches the failure. This may result in unexpected behaviour, so follow this approach only if your use case merits it.
:::
```

Author: It's only unexpected behavior if the developer isn't aware of the implications. When that's understood, it's a matter of "knowing what you're doing", depending on the situation.
